File: backend/src/main/java/top/vuhe/entity/equipment/bo/StateBO.java
package top.vuhe.entity.equipment.bo;

import lombok.Data;
import top.vuhe.entity.equipment.ElectricInfo;
import top.vuhe.entity.equipment.dto.StatusDTO;
import top.vuhe.entity.equipment.dto.ThresholdDTO;

import java.util.*;

/**
 * @author zhuhe
 */
@Data
public class StateBO implements ElectricInfo {
    private static final int ROW_LEN = 2;
    private final List<StatusDTO> status;

    public StateBO(List<Byte> bytes) {
        if (bytes.size() % ROW_LEN != 0) {
            throw new IllegalArgumentException("读取byte时出错");
        }
        Iterator<Byte> it = bytes.iterator();
        List<StatusDTO> statusList = new LinkedList<>();
        // 存储实时值
        int length = bytes.size() / ROW_LEN;
        for (int i = 0; i < length; i++) {
            int channel = it.next();
            int value = it.next();
            statusList.add(new StatusDTO(channel, value));
        }
        status = Collections.unmodifiableList(statusList);
    }

    @Override
    public Map<Integer, Double> getRealTimeDTO() {
        return null;
    }

    @Override
    public List<StatusDTO> getStatusDTO() {
        return status;
    }

    @Override
    public List<ThresholdDTO> getThresholdDTO() {
        return null;
    }
}

File: backend-kt/src/main/kotlin/top/vuhe/sw/entity/equipment/bo/StateBO.kt
package top.vuhe.sw.entity.equipment.bo

import top.vuhe.sw.entity.equipment.ElectricInfo
import top.vuhe.sw.entity.equipment.dto.RealTimeDTO
import top.vuhe.sw.entity.equipment.dto.StatusDTO
import top.vuhe.sw.entity.equipment.dto.ThresholdDTO

class StateBO(bytes: List<Byte>) : ElectricInfo {
    companion object {
        private const val ROW_LEN = 2
        private val hasTowValue = setOf(
            37, 39, 41, 43, 45, 47, 49, 51, 53, 55, 57, 59
        )
    }

    private val status: StatusDTO

    init {
        require(bytes.size % ROW_LEN == 0) { "读取byte时出错" }
        val it = bytes.iterator()
        val statusMap: MutableMap<Int, Int> = HashMap(52)
        // 存储实时值
        val length = bytes.size / ROW_LEN
        for (i in 0 until length) {
            val channel = it.next().toInt() and 0xFF
            val value = it.next().toInt() and 0xFF
            statusMap[channel]= value
        }
        // 合并高低位
        for (ch in hasTowValue) {
            if (statusMap[ch] != null) {
                statusMap.remove(ch + 1)
            }
        }
        status = statusMap
    }

    override fun getRealTimeDTO(): RealTimeDTO? {
        return null
    }

    override fun getStatusDTO(): StatusDTO {
        return status
    }

    override fun getThresholdDTO(): ThresholdDTO? {
        return null
    }
}